x-pack/plugin/sql/src/main/java/org/elasticsearch/xpack/sql/plan/logical/command/sys/SysGeometryColumns.java

```diff
@@ -0,0 +1,144 @@
+    protected NodeInfo<SysGeometryColumns> info() {
+        return NodeInfo.create(this, SysGeometryColumns::new, catalog, indexPattern, columnPattern);
+    }
+
+    @Override
+    public List<Attribute> output() {
+        return asList(keyword("F_TABLE_CATALOG"),        // Cluster name
+            keyword("F_TABLE_SCHEMA"),         // Always null - not supported
+            keyword("F_TABLE_NAME"),           // Index name
+            keyword("F_GEOMETRY_COLUMN"),      // Geo field name
+            field("STORAGE_TYPE", INTEGER),    // Always 1 for now
+            field("GEOMETRY_TYPE", INTEGER),   // 0 - for geoshapes or 1 - for geopoints
+            field("COORD_DIMENSION", INTEGER), // Always 2
+            field("SRID", INTEGER),            // Always 0 (the ID of the Spatial Reference System)
+            keyword("TYPE")                    // GEOMETRY for geoshapes, POINT for geopoints
+        );
+    }
+
+    @Override
+    public void execute(SqlSession session, ActionListener<SchemaRowSet> listener) {
+        String cluster = session.indexResolver().clusterName();
+
+        // bail-out early if the catalog is present but differs
+        if (Strings.hasText(catalog) && !cluster.equals(catalog)) {
+            listener.onResponse(Rows.empty(output()));
+            return;
+        }
+
+        String index = indexPattern != null ? indexPattern.asIndexNameWildcard() : "*";
+        String regex = indexPattern != null ? indexPattern.asJavaRegex() : null;
+
+        Pattern columnMatcher = columnPattern != null ? Pattern.compile(columnPattern.asJavaRegex()) : null;
+
+        session.indexResolver().resolveAsSeparateMappings(index, regex, ActionListener.wrap(esIndices -> {
+            List<List<?>> rows = new ArrayList<>();
+            for (EsIndex esIndex : esIndices) {
+                fillInRows(cluster, esIndex.name(), esIndex.mapping(), null, rows, columnMatcher);
+            }
+
+            listener.onResponse(Rows.of(output(), rows));
+        }, listener::onFailure));
+    }
+
+    static void fillInRows(String clusterName, String indexName, Map<String, EsField> mapping, String prefix, List<List<?>> rows,
+                           Pattern columnMatcher) {
+        int pos = 0;
+        for (Map.Entry<String, EsField> entry : mapping.entrySet()) {
+            pos++; // JDBC is 1-based so we start with 1 here
+
+            String name = entry.getKey();
+            name = prefix != null ? prefix + "." + name : name;
+            EsField field = entry.getValue();
+            DataType type = field.getDataType();
+            if ((columnMatcher == null || columnMatcher.matcher(name).matches()) && type == GEO_SHAPE) {
+                rows.add(asList(clusterName,  // catalog
+                    null,                 // schema is not supported
+                    indexName,            // table name
+                    name,                 // column name
+                    1,                    // storage type
```

Comment:
Can you please put these constants into something like `GeoUtils` so `DataTypes`? I went with a similar approach in #30418 for readability and maintenance.